ifsbench/application.py

```diff
@@ -0,0 +1,149 @@
+# (C) Copyright 2020- ECMWF.
+# This software is licensed under the terms of the Apache Licence Version 2.0
+# which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+# In applying this licence, ECMWF does not waive the privileges and immunities
+# granted to it by virtue of its status as an intergovernmental organisation
+# nor does it submit to any jurisdiction.
+
+from abc import ABC, abstractmethod
+
+__all__ = ['Application', 'DefaultApplication']
+
+class Application(ABC):
+    """
+    Base class for applications that can be launched.
+    """
+
+    @abstractmethod
+    def get_data_handlers(self, run_dir, job):
+        """
+        Return necessary data handlers.
+
+        Return all application-specific handler objects that must be run before
+        launching the executable.
+
+        Parameters
+        ----------
+        run_dir: :any:`pathlib.Path`
+            The working directory from where the executable will be launched.
+        job: Job
+            The parallel job setup.
+
+        Returns
+        -------
+        list[DataHandler]:
+            List of DataHandler objects.
+        """
+        return NotImplemented
```

Comment:
I don't think the explicit return statements are required. A derived class that doesn't implement these methods can't be instantiated, and therefore this line can never be reached, [see](https://docs.python.org/3/library/abc.html#abc.abstractmethod):
> Using this decorator requires that the class’s metaclass is [ABCMeta](https://docs.python.org/3/library/abc.html#abc.ABCMeta) or is derived from it. A class that has a metaclass derived from ABCMeta cannot be instantiated unless all of its abstract methods and properties are overridden.